util/rateLimit.go

```diff
@@ -0,0 +1,470 @@
+	q <- capacity{}
+}
+
+func (q capacityQueue) blockingConsume() {
+	<-q
+}
+
+func (q capacityQueue) consume(cm *CongestionManager) (ErlCapacityGuard, error) {
+	select {
+	case <-q:
+		return ErlCapacityGuard{
+			cq: q,
+			cm: cm,
+		}, nil
+	default:
+		return ErlCapacityGuard{}, fmt.Errorf("could not consume capacity from capacityQueue: %v", q)
+	}
+}
+
+// NewElasticRateLimiter creates an ElasticRateLimiter and initializes maps
+func NewElasticRateLimiter(
+	maxCapacity int,
+	reservedCapacity int,
+	cm CongestionManager,
+	nocapCount *metrics.Counter,
+	conmanCount *metrics.Counter) *ElasticRateLimiter {
+	ret := ElasticRateLimiter{
+		MaxCapacity:              maxCapacity,
+		CapacityPerReservation:   reservedCapacity,
+		capacityByClient:         map[ErlClient]capacityQueue{},
+		cm:                       cm,
+		sharedCapacity:           capacityQueue(make(chan capacity, maxCapacity)),
+		noCapacityCounter:        nocapCount,
+		congestionControlCounter: conmanCount,
+	}
+	// fill the sharedCapacity
+	for i := 0; i < maxCapacity; i++ {
+		ret.sharedCapacity.blockingRelease()
+	}
+	return &ret
+}
+
+// EnableCongestionControl turns on the flag that the ERL uses to check with its CongestionManager
+func (erl *ElasticRateLimiter) EnableCongestionControl() {
+	erl.clientLock.Lock()
+	erl.enableCM = true
+	erl.clientLock.Unlock()
+}
+
+// DisableCongestionControl turns off the flag that the ERL uses to check with its CongestionManager
+func (erl *ElasticRateLimiter) DisableCongestionControl() {
+	erl.clientLock.Lock()
+	erl.enableCM = false
+	erl.clientLock.Unlock()
+}
+
+// ConsumeCapacity will dispense one capacity from either the resource's reservedCapacity,
+// and will return a guard who can return capacity when the client is ready
+func (erl *ElasticRateLimiter) ConsumeCapacity(c ErlClient) (ErlCapacityGuard, error) {
+	var q capacityQueue
+	var err error
+	var exists bool
+	var enableCM bool
+	// get the client's queue
+	erl.clientLock.RLock()
+	q, exists = erl.capacityByClient[c]
+	enableCM = erl.enableCM
+	erl.clientLock.RUnlock()
+
+	// Step 0: Check for, and create a capacity reservation if needed
+	if !exists {
+		q, err = erl.openReservation(c)
+		if err != nil {
+			return ErlCapacityGuard{}, err
+		}
+		// if the client has been given a new reservation, make sure it cleans up OnClose
+		c.OnClose(func() { erl.closeReservation(c) })
+
+		// if this reservation is newly created, directly (blocking) take a capacity
+		q.blockingConsume()
+		fmt.Println(len(q))
+		return ErlCapacityGuard{cq: q, cm: &erl.cm}, nil
+	}
+
+	// Step 1: Attempt consumption from the reserved queue
+	cg, err := q.consume(&erl.cm)
+	if err == nil {
+		if erl.cm != nil {
+			erl.cm.Consumed(c, time.Now()) // notify the congestion manager that this client consumed from this queue
+		}
+		return cg, nil
+	}
+	// Step 2: Potentially gate shared queue access if the congestion manager disallows it
+	if erl.cm != nil &&
+		enableCM &&
+		erl.cm.ShouldDrop(c) {
```

Comment:
ConsumeCapacity() is called on each new message, and  ShouldDrop does not return immediately since it can block on reading response, that can be delayed by 1) other channels are ready in `run()` worker 2) `run()` has a loop iterating all clients, not sure how big it could be.

Reply:
Yes, at an earlier revision the ShouldDrop queries were handled as their loop, so that they could be all addressed on every pass. I merged it in for cleanliness, but this is a good point regarding fast responses.